src/openrct2-ui/drawing/engines/opengl/TextureCache.cpp

```diff
@@ -23,35 +23,69 @@
 
 #include <openrct2/drawing/drawing.h>
 
+constexpr uint32 UNUSED_INDEX = 0xFFFFFFFF;
+
+TextureCache::TextureCache()
+{
+    std::fill(_indexMap.begin(), _indexMap.end(), UNUSED_INDEX);
+}
+
 TextureCache::~TextureCache()
 {
     FreeTextures();
 }
 
 void TextureCache::InvalidateImage(uint32 image)
 {
-    auto kvp = _imageTextureMap.find(image);
-    if (kvp != _imageTextureMap.end())
+    uint32 index = _indexMap[image];
+    if (index == UNUSED_INDEX)
+        return;
+
+    CachedTextureInfo& elem = _textureCache.at(index);
+
+    _atlases[elem.index].Free(elem);
+    _indexMap[image] = UNUSED_INDEX;
+
+    if (index == _textureCache.size() - 1)
     {
-        _atlases[kvp->second.index].Free(kvp->second);
-        _imageTextureMap.erase(kvp);
+        // Last element can be popped back.
+        _textureCache.pop_back();
+    }
+    else
+    {
+        // Swap last element with element to erase and then pop back.
+        CachedTextureInfo& last = _textureCache.back();
+
+        // Move last to current.
+        elem = last;
+
+        // Change index for moved element.
+        _indexMap[last.image] = index;
+
+        _textureCache.pop_back();
     }
 }
 
 const CachedTextureInfo* TextureCache::GetOrLoadImageTexture(uint32 image)
 {
     image &= 0x7FFFF;
 
-    auto kvp = _imageTextureMap.find(image);
-    if (kvp != _imageTextureMap.end())
+    uint32 index = _indexMap[image];
+    if (index != UNUSED_INDEX)
     {
-        return &kvp->second;
+        return &_textureCache[index];
     }
 
-    auto cacheInfo = LoadImageTexture(image);
-    auto cacheItr = _imageTextureMap.insert(std::make_pair(image, cacheInfo));
+    index = (uint32)_textureCache.size();
+
+    CachedTextureInfo info = LoadImageTexture(image);
+    info.image = image;
+
+    _textureCache.push_back(info);
+    
+    _indexMap[image] = index;
 
-    return &(cacheItr.first->second);
+    return &_textureCache.at(index);
```

Comment:
You can't return it like this, if you're calling `push_back`. The returned pointer outlives the pointee whenever vector is out of capacity.

Reply:
This won't be happening with the current code since all fields are copied of into the queue.

Reply:
I suppose I could turn it to a reference to clarify the lifetime.